shell/platform/embedder/embedder.cc

```diff
@@ -1249,6 +1249,158 @@ void PopulateAOTSnapshotMappingCallbacks(
+    const flutter::CustomAccessibilityAction& action) {
+  return {
+      sizeof(FlutterSemanticsCustomAction),
+      action.id,
+      static_cast<FlutterSemanticsAction>(action.overrideId),
+      action.label.c_str(),
+      action.hint.c_str(),
+  };
+}
+
+// Creates a callback that receives semantic updates from the engine
+// and notifies the embedder's callback(s). Returns null if the embedder
+// did not register any callbacks.
+flutter::PlatformViewEmbedder::UpdateSemanticsCallback
+CreateEmbedderSemanticsUpdateCallback(const FlutterProjectArgs* args,
+                                      void* user_data) {
+  // The embedder can register the new callback, or the old callbacks, or
+  // nothing at all. Handle the case where the embedder registered the 'new'
+  // callback.
+  if (SAFE_ACCESS(args, update_semantics_callback, nullptr) != nullptr) {
+    return [ptr = args->update_semantics_callback, user_data](
+               const flutter::SemanticsNodeUpdates& nodes,
+               const flutter::CustomAccessibilityActionUpdates& actions) {
+      std::vector<FlutterSemanticsNode> embedder_nodes;
+      for (const auto& value : nodes) {
+        embedder_nodes.push_back(CreateEmbedderSemanticsNode(value.second));
+      }
+
+      std::vector<FlutterSemanticsCustomAction> embedder_custom_actions;
+      for (const auto& value : actions) {
+        embedder_custom_actions.push_back(
+            CreateEmbedderSemanticsCustomAction(value.second));
+      }
+
+      FlutterSemanticsUpdate update{
+          .struct_size = sizeof(FlutterSemanticsUpdate),
+          .nodes_count = embedder_nodes.size(),
+          .nodes = embedder_nodes.data(),
+          .custom_actions_count = embedder_custom_actions.size(),
+          .custom_actions = embedder_custom_actions.data(),
+      };
+
+      ptr(&update, user_data);
+    };
+  }
+
+  FlutterUpdateSemanticsNodeCallback update_semantics_node_callback = nullptr;
+  if (SAFE_ACCESS(args, update_semantics_node_callback, nullptr) != nullptr) {
+    update_semantics_node_callback = args->update_semantics_node_callback;
+  }
+
+  FlutterUpdateSemanticsCustomActionCallback
+      update_semantics_custom_action_callback = nullptr;
+  if (SAFE_ACCESS(args, update_semantics_custom_action_callback, nullptr) !=
+      nullptr) {
+    update_semantics_custom_action_callback =
+        args->update_semantics_custom_action_callback;
+  }
+
+  // Handle the case where the embedder registered no callbacks.
+  if (update_semantics_node_callback == nullptr &&
+      update_semantics_custom_action_callback == nullptr) {
+    return nullptr;
+  }
+
+  // Handle the case where the embedder registered 'old' callbacks.
+  return [update_semantics_node_callback,
+          update_semantics_custom_action_callback,
+          user_data](const flutter::SemanticsNodeUpdates& nodes,
+                     const flutter::CustomAccessibilityActionUpdates& actions) {
```

Comment:
Below is the original logic to call the old embedder callbacks: 

https://github.com/flutter/engine/blob/243019f734a29acf13c87dfbf9dc12ee03a2d54d/shell/platform/embedder/embedder.cc#L1425-L1507